### PR TITLE
Core/Spells: fix Priest spell Penance and rewrite Power Word:Shield

### LINKED ISSUE #18520: Core/Spell: Penance should not heal until talent is set
**Description:**  The priest spell "penance" has more functions than the blizzlike spell once learned (info from starter edition account)  

**Current behaviour:**  After learning "penance", the spell has the capabilities of healing allied units and damaging the enemy units  

**Expected behaviour:**  This spell at the beginning when you first learn it must only be able of damaging enemy units and only when you apply the required talent this spell should heal allied units.

**Steps to reproduce the problem:**  
1- Create a priest character  
2- Obtain level 10  
3- Use "Penance" on an allied unit and an enemy unit.

**Branch(es): 3.3.5 / master (Tell us which branch(es) this issue affects.)**   Master  

**TC rev. hash/commit:**  bc758991140e961ceac46b17b766b0d54b8129fb  

**TDB version:**  703.00 + updates  

**Operating system:** Windows 10

### PR DESCRIPTION
**Pull request by HannibalRoG**
[//]: # (***************************)
[//]: # (** FILL IN THIS TEMPLATE **)
[//]: # (***************************)

**Changes proposed:**

-  Correct implement of talent Penitent for Spell Penance and re-write Power Word Shield.

**Target branch(es):** 3.3.5/master

- master

**Issues addressed:** Closes #  (insert issue tracker number)
Closes #18520

**Tests performed:** (Does it build, tested in-game, etc.)
Compiled and tested in game.

### CODE REVIEW

**Comment by lorac2k14** [comment]
whats going on with your spacing it is all over the place? make sure you are following [C++ Development Standards](https://trinitycore.atlassian.net/wiki/pages/viewpage.action?pageId=2130103)

**Comment by Keader** [comment]
use 4 spaces instead tabs

File: src/server/scripts/Spells/spell_priest.cpp [hunk 1]
```diff
@@ -1241,6 +1182,106 @@ class spell_pri_vampiric_touch : public SpellScriptLoader
...
+            int32 n = dmgInfo.GetDamage() + absorbAmount;
+            int32 m = aurEff->GetBase()->GetEffect(0)->CalculateAmount(caster);
+            if (n > m) return;
+            absorbAmount += (int32)dmgInfo.GetDamage();
+        }
+
+        void HandleOnRemove(const AuraEffect* aurEff, AuraEffectHandleModes mode)
+        {
+            Player* caster = GetCaster()->ToPlayer();
+            int32 max = aurEff->GetBase()->GetMaxDuration();
```
Comment (tkrokli) [inline src/server/scripts/Spells/spell_priest.cpp]:
```
src/server/scripts/Spells/spell_priest.cpp:1254:19: fatal error: 
      unused variable 'max' [-Wunused-variable]
            int32 max = aurEff->GetBase()->GetMaxDuration();
                  ^
```

File: src/server/scripts/Spells/spell_priest.cpp [hunk 2]
```diff
@@ -1241,6 +1182,106 @@ class spell_pri_vampiric_touch : public SpellScriptLoader
...
+        SPELL_PRIEST_BODY_AND_SOUL_ACTIVE       = 65081,
+        SPELL_PRIEST_SHIELD_DISCIPLINE_PASSIVE  = 197045,
+        SPELL_PRIEST_SHIELD_DISCIPLINE_ENERGIZE = 47755,
+        SPELL_PRIEST_STRENGHT_OF_SOUL           = 197535,
+        SPELL_PRIEST_STRENGHT_OF_SOUL_EFFECT    = 197548,
+        SPELL_PRIEST_RENEWED_HOPE               = 197469,
+        SPELL_PRIEST_RENEWED_HOPE_EFFECT        = 197470,
+        SPELL_PRIEST_VOID_SHIELD                = 199144,
+        SPELL_PRIEST_VOID_SHIELD_EFFECT         = 199145
+    };
```
Comment (P-Kito) [inline src/server/scripts/Spells/spell_priest.cpp]: missing line

File: src/server/scripts/Spells/spell_priest.cpp [hunk 3]
```diff
@@ -1241,6 +1182,106 @@ class spell_pri_vampiric_touch : public SpellScriptLoader
...
+    {
+
+        PrepareSpellScript(spell_pri_power_word_shield_SpellScript);
+
+        void HandleAfterCast()
+        {
+            Player* caster = GetCaster()->ToPlayer();
+
+            if (caster->HasAura(SPELL_PRIEST_BODY_AND_SOUL))
+                caster->CastSpell(caster, SPELL_PRIEST_BODY_AND_SOUL_ACTIVE,true);
```
Comment (P-Kito) [inline src/server/scripts/Spells/spell_priest.cpp]: Codestyle ", true"

File: src/server/scripts/Spells/spell_priest.cpp [hunk 4]
```diff
@@ -1241,6 +1182,106 @@ class spell_pri_vampiric_touch : public SpellScriptLoader
...
+        SPELL_PRIEST_STRENGHT_OF_SOUL           = 197535,
+        SPELL_PRIEST_STRENGHT_OF_SOUL_EFFECT    = 197548,
+        SPELL_PRIEST_RENEWED_HOPE               = 197469,
+        SPELL_PRIEST_RENEWED_HOPE_EFFECT        = 197470,
+        SPELL_PRIEST_VOID_SHIELD                = 199144,
+        SPELL_PRIEST_VOID_SHIELD_EFFECT         = 199145
+    };
+    class spell_pri_power_word_shield_SpellScript : public SpellScript
+    {
+
```
Comment (P-Kito) [inline src/server/scripts/Spells/spell_priest.cpp]: remove this line :P

File: src/server/scripts/Spells/spell_priest.cpp [hunk 5]
```diff
@@ -1241,6 +1182,106 @@ class spell_pri_vampiric_touch : public SpellScriptLoader
...
+        void CalculateAmount(const AuraEffect* /*auraEffect*/, int32& amount, bool& /*canBeRecalculated*/)
+        {
+            Player* caster = GetCaster()->ToPlayer();
+            int32 sp = caster->SpellBaseDamageBonusDone(SPELL_SCHOOL_MASK_HOLY);
+            int32 versadmg = caster->GetRatingBonusValue(CombatRating::CR_VERSATILITY_DAMAGE_DONE);
+            int32 mastery = caster->GetRatingBonusValue(CombatRating::CR_MASTERY) + 1;
+            amount = (int32)((sp * 5.5f) + mastery) * (1 + versadmg);
+        }
+        
+        void Absorb(AuraEffect* aurEff, DamageInfo & dmgInfo, uint32 & absorbAmount)
```
Comment (P-Kito) [inline src/server/scripts/Spells/spell_priest.cpp]: this function is not crash safe, make sure GetEffect is checked and ToPlayer is checked, also rename "n" and "m" to something easier to understand. 
Another suggestion:

```c++
if (n <= m)
    absorbAmount += (int32)dmgInfo.GetDamage();
```

**Comment by P-Kito** [comment]
Fix the codestyle, there are alot of unclear variable names, missing spaces or too much space.

I am not going to comment them all, but please double check the code and fix the codestyle.
(I already commented some)

File: sql/updates/world/master/2017_02_28_19033_world.sql [hunk 6]
```diff
@@ -0,0 +1,5 @@
+-- Class: Priest
+-- Spec: All
+-- Spell: Power Word:Shield
+DELETE FROM `spell_script_names` WHERE ScriptName = 'spell_pri_power_word_shield';
+INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES ('17', 'spell_pri_power_word_shield');
```
Comment (P-Kito) [inline sql/updates/world/master/2017_02_28_19033_world.sql]: Missing new line at the end :-)

Comment (Aokromes) [inline sql/updates/world/master/2017_02_28_19033_world.sql]: nope, no "no newline" mark on diff.

Comment (P-Kito) [inline sql/updates/world/master/2017_02_28_19033_world.sql]: My bad, didn't look like it :P

Comment (tkrokli) [inline sql/updates/world/master/2017_02_28_19033_world.sql]: I fixed it in my commit 98bb86c162d354a512725945f37144921a22befc

**Comment by Aokromes** [comment]
/home/travis/build/TrinityCore/TrinityCore/src/server/scripts/Spells/spell_priest.cpp:1254:19: fatal error: 
      unused variable 'max' [-Wunused-variable]
            int32 max = aurEff->GetBase()->GetMaxDuration();
                  ^

**Comment by Edreisantafe** [comment]
I am taking note, lol.

**Comment by Edreisantafe** [comment]
Well, at last it seems Travis is happy.

**Comment by Edreisantafe** [comment]
Well, at last it seems Travis is happy.

**Comment by Edreisantafe** [comment]
Compiled and tested, I think it's Blizzlike. Any corrections?

**Comment by tkrokli** [comment]
Code style: add 1 CR/LF at the end of file (EOF) of `spell_priest.cpp` - your typing cursor should always be able to move 1 line below the last line of code in any of the files.

File: sql/updates/world/master/2017_02_28_19033_world.sql [hunk 7]
```diff
@@ -0,0 +1,5 @@
+-- Class: Priest
+-- Spec: All
+-- Spell: Power Word:Shield
+DELETE FROM `spell_script_names` WHERE ScriptName = 'spell_pri_power_word_shield';
```
Comment (offl) [inline sql/updates/world/master/2017_02_28_19033_world.sql]: Always use the grave accent for table or column names.
```sql
DELETE FROM `spell_script_names` WHERE `ScriptName` = 'spell_pri_power_word_shield';
```

File: sql/updates/world/master/2017_02_28_19033_world.sql (hunk 6 above)
Comment (offl) [inline sql/updates/world/master/2017_02_28_19033_world.sql]: Never use apostrophes for numbers.
```sql
INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES (17, 'spell_pri_power_word_shield');
```

File: src/server/scripts/Spells/spell_priest.cpp [hunk 8]
```diff
@@ -755,6 +757,7 @@ class spell_pri_pain_and_suffering_proc : public SpellScriptLoader
                         aur->SetDamage(caster->SpellDamageBonusDone(target, aur->GetSpellInfo(), damage, DOT, aur->GetSpellEffectInfo()) * aur->GetDonePct());
                         aur->CalculatePeriodic(caster, false, false);
                         aur->GetBase()->RefreshDuration();
+
```
Comment (offl) [inline src/server/scripts/Spells/spell_priest.cpp]: Excess line.

File: src/server/scripts/Spells/spell_priest.cpp [hunk 9]
```diff
@@ -1269,4 +1321,4 @@ void AddSC_priest_spell_scripts()
     new spell_pri_vampiric_embrace();
     new spell_pri_vampiric_embrace_target();
     new spell_pri_vampiric_touch();
-}
+}
```
Comment (offl) [inline src/server/scripts/Spells/spell_priest.cpp]: "No newline at end of file"

File: src/server/scripts/Spells/spell_priest.cpp [hunk 10]
```diff
@@ -1241,6 +1181,117 @@ class spell_pri_vampiric_touch : public SpellScriptLoader
...
+            }
+        }
+
+        void Register() override
+        {
+            AfterCast += SpellCastFn(spell_pri_power_word_shield_SpellScript::HandleAfterCast);
+        }
+    };
+
+    class spell_pri_power_word_shield_AuraScript :public AuraScript
```
Comment (offl) [inline src/server/scripts/Spells/spell_priest.cpp]: Missing space.
```cpp
    class spell_pri_power_word_shield_AuraScript : public AuraScript
```

File: sql/updates/world/master/2017_02_28_19033_world.sql [hunk 11]
```diff
@@ -0,0 +1,5 @@
+-- Class: Priest
+-- Spec: All
+-- Spell: Power Word:Shield
+DELETE FROM `spell_script_names` WHERE `ScriptName` = 'spell_pri_power_word_shield';
+INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES ('17', 'spell_pri_power_word_shield');
```
Comment (offl) [inline sql/updates/world/master/2017_02_28_19033_world.sql]: Again.
```sql
INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES (17, 'spell_pri_power_word_shield');
```

**Comment by wonder2011** [comment]
I test this code and is working fine.

**Comment by Edreisantafe** [comment]
Doing a research with my starter edition account I found two problems that are not contemplated in the script, one is that the amount of absorption of the shield is not so high, that is, the formula seems not correct, and two, the aura Updates every time the shield absorbs damage, this is more of a visual problem.
I will review the code and add these two problems, I do not know if they should wait for them to be fixed to include this or add these first and then the arrangements.
If you could give me some suggestion as to the formula I would be grateful.

**Comment by Edreisantafe** [comment]
Bueno, más investigaciones han arrojado lo siguiente:
La fórmula parece estar bastante cercana a la realidad, comparando las imágenes retail 01.jpg y trinity 01.jpg vemos poca diferencia, de hecho en el script está por debajo.
retail 01
trinity 01
Además con un objeto con 2 de versatilidad los números son bastante similares.
retail 02
trinity 02
Lo que sucede es que cuando tenemos, por ejemplo, 8 de versatilidad el número aumenta drásticamente 771 puntos, este objeto "Leaving Root" no debiera tener tanta versatilidad en nivel 20.
Las pruebas iniciales me habían mostrado grandes números de absorción debido a la cantidad de versatilidad que presentaba el PJ.
Sólo queda implementar que se actualice el aura con el número restante de absorción de daño.
Nótece el bug de regeneración de maná, en el retail es de 80 puntos y en trinity 200 cada 5s, esto hace que el maná del sacerdote sea infinito.

**Comment by HannibalRoG** [comment]
Any news with this PR?

**Comment by tkrokli** [comment]
If there are no objections or mentions of things to change, we could merge it before Monday.

**Comment by tkrokli** [comment]
Thank you for the PR, your work is appreciated. 👍

**Comment by wonder2011** [comment]
The PVP Talents are implemented ?

**Pull request merged**